### PR TITLE
Add a ?Sized bound for box_error conversion

### PR DESCRIPTION
**Pull request by nagisa**
`Box<dyn Error + …>` has lately been gaining in popularity as the error
type in the Rust ecosystem (esp. between people who don’t need anything
more than just type erasure). This type, however is not as full-featured
as some alternatives in the crates ecosystem. Namely, it does not itself
implement `std::error::Error` because the relevant implementation is
lacking the `T: ?Sized` bound relaxation.

Remembering that for trait to be object safe (and thus making it
possible to make a trait object of), its methods must be able to accept
a `Self: ?Sized`. `Self = T` in this instance, and therefore this
relaxation of bounds should not have any stability implications.

TL;DR: This allows code like the following to work:

```rust
fn main() {
    let x: Box<dyn std::error::Error> = std::io::Error::last_os_error().into();
    fn use_error<E: std::error::Error>(x: E) {}
    use_error(x);
}
```

### CODE REVIEW

**Comment by nagisa** [comment]
Hmm, looks like this is is conflicting with 

```
impl<'a, E: Error + 'a> From<E> for Box<dyn Error + 'a>
impl<'a, E: Error + Send + Sync + 'a> From<E> for Box<dyn Error + Send + Sync + 'a>
```

and we cannot remove those either :(
